internal/hcl/modules/registry.go

```diff
@@ -0,0 +1,250 @@
+package modules
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	goversion "github.com/hashicorp/go-version"
+	"github.com/pkg/errors"
+)
+
+var defaultRegistryHost = "registry.terraform.io"
+
+// validRegistryName is a regexp that matches valid registry identifier for namespaces, module names and targets
+var validRegistryName = regexp.MustCompile("^[0-9A-Za-z-_]+$")
+
+// RegistryLookupResult is returned when looking up the module to check if it exists in the registry
+// and has a matching version
+type RegistryLookupResult struct {
+	Source      string
+	Version     string
+	DownloadURL string
+}
+
+// RegistryLoader is a loader that can lookup modules from a Terraform Registry and download them to the given destination
+type RegistryLoader struct {
+	packageFetcher *PackageFetcher
+	dest           string
+}
+
+// NewRegistryLoader constructs a registry loader
+func NewRegistryLoader(packageFetcher *PackageFetcher, dest string) *RegistryLoader {
+	return &RegistryLoader{
+		packageFetcher: packageFetcher,
+		dest:           dest,
+	}
+}
```

Comment:
I'd change the `dest` to move to an input of the `downloadModule` method. Then we can have `RegistryLoader` as part of the top level `Loader` and not have to be constructed in a method.

Reply:
e.g.:

```go
        fetcher :=  NewPackageFetcher()
	return &ModuleLoader{
		Path:           path,
		cache:          NewCache(),
		packageFetcher: packageFetcher,
		registryLoader: RegistryLoader{packageFetcher: packageFetcher}
	}
```